core/src/main/java/cucumber/runtime/Shellwords.java

```diff
@@ -6,27 +6,43 @@
 import java.util.regex.Pattern;
 
 public class Shellwords {
-    private static final Pattern SHELLWORDS_PATTERN = Pattern.compile("[^\\s'\"]+|[']([^']*)[']|[\"]([^\"]*)[\"]");
+
+    private static final Pattern SHELLWORDS_PATTERN;
+
+    static {
+        final String SHELLWORDS_FORMAT = "\\s*(?>([^\\s\\\\\\'\\\"]+)|'([^\\']*)'|\"((?:[^\\\"\\\\]|\\\\.)*)\"|(\\\\.?)|(\\S))(\\s|\\z)?";
```

Comment:
In one of your previous commits you split this up into several variables. This made the purpose of each capture group clear. That was a good idea.

Reply:
You can also name capturing groups using `(?<name>X)` where X is the regex.

Reply:
I'm not 100% sure which each of these capture groups does...
By debugging I've worked out the 2nd highest capture group is the capture group that should be used, so using names would make that harder instead of easier. I'm also assuming I can removed the last capture group but when I do the highest capture group isn't always the one you want...

Reply:
The algorithm used in Ruby isn't trivial. I'll try to explain it line by line:

https://github.com/sj26/ruby-1.9.3-p0/blob/master/lib/shellwords.rb#L32-L34
```rb
words = []
field = ''
```
Our goal is to out put an array of words (strings separated by whitespace, unless quoted). To do this we are going to search through the string. We'll put the intermediate results that we find into field. When we encounter a separator we'll know that we've found a complete word.

https://github.com/sj26/ruby-1.9.3-p0/blob/master/lib/shellwords.rb#L35-L36
```rb
    line.scan(/\G\s*(?>([^\s\\\'\"]+)|'([^\']*)'|"((?:[^\"\\]|\\.)*)"|(\\.?)|(\S))(\s|\z)?/m) do
|word, sq, dq, esc, garbage, sep|
```
This searches the line for any bits that match the regular expression. The results is an array that contains the captured values of each capture group. The elements of this array are unpacked into the variables word, sq, dq, esc, garbage and sep.

https://github.com/sj26/ruby-1.9.3-p0/blob/master/lib/shellwords.rb#L37
```rb
 raise ArgumentError, "Unmatched double quote: #{line.inspect}" if garbage
```
If the garbage capture group capture anything it throws an exception.

https://github.com/sj26/ruby-1.9.3-p0/blob/master/lib/shellwords.rb#L38-L41
```rb
      field << (word || sq || (dq || esc).gsub(/\\(?=.)/, ''))
      if sep
        words << field
        field = ''
```
Then the fields word, single quote, double quote and escaped are appended to the current field we are building. The latter two have any escape character removed. Note that the separator character is not appended.

If there is a separator or line ending  we are done building the current field and put it in the list of fields.

Reply:
The underlying principle is called [Lexical Analysis](https://en.wikipedia.org/wiki/Lexical_analysis). And normally you'd write a state machine that follows the lexical grammar. In this case the Ruby implementation is being clever by expressing that state machine as a regular expression.

That said. We shouldn't be in the business of being clever. It creates unreadable code.

Looking through the Java API I found the [StreamTokenizer](https://docs.oracle.com/javase/8/docs/api/java/io/StreamTokenizer.html). You may find it useful.